framework-cloud/framework-openshift/src/main/java/org/kie/cloud/openshift/log/InstancesLogCollectorRunnable.java

```diff
@@ -0,0 +1,103 @@
+package org.kie.cloud.openshift.log;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.kie.cloud.api.deployment.Instance;
+import org.kie.cloud.common.logs.InstanceLogUtil;
+import org.kie.cloud.openshift.deployment.OpenShiftInstance;
+import org.kie.cloud.openshift.resource.Project;
+import org.kie.cloud.openshift.scenario.OpenShiftScenario;
+import org.kie.cloud.openshift.util.OpenshiftInstanceUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InstancesLogCollectorRunnable implements Runnable {
+
+    private static final Logger logger = LoggerFactory.getLogger(OpenShiftScenario.class);
+
+    private static final Integer DEFAULT_OBERVABLE_BUFFER_IN_SECONDS = 5;
+
+    private Project project;
+    private String logFolderName;
+
+    private ExecutorService executorService = Executors.newCachedThreadPool();
+    private Set<OpenShiftInstance> observedInstances = Collections.synchronizedSet(new HashSet<>());
+
+    public InstancesLogCollectorRunnable(Project project, String logFolderName) {
+        super();
+        this.project = project;
+        this.logFolderName = logFolderName;
+    }
+
+    @Override
+    public void run() {
+        // Check for new instances and observe on them
+        OpenshiftInstanceUtil.getAllInstances(project.getOpenShift(), project.getName()).stream()
+                             // Filter non observed instances
+                             .filter(instance -> !isInstanceObserved(instance))
+                             // Observe instance logs
+                             .forEach(this::observeInstanceLog);
+
+        // Check if instances are still existing
+        // Calling get pod on a no more existing pod will cause the "obervePodLog" to complete ...
+    }
+
+    public void closeAndFlushRemainingInstanceCollectors() {
+        // Keep instances list before stopping everything
+        List<OpenShiftInstance> instances = new ArrayList<>(observedInstances);
+        // Stop all observations
+        executorService.shutdownNow();
+        // Flush all remaining instances (which were still observed)
+        instances.forEach(this::flushInstanceLogs);
+    }
+
+    private void observeInstanceLog(OpenShiftInstance instance) {
+        Future<?> future = executorService.submit(() -> {
+            instance.observeLogs().buffer(DEFAULT_OBERVABLE_BUFFER_IN_SECONDS, TimeUnit.SECONDS)
+                    .subscribe(logLines -> instanceLogLines(instance, logLines), error -> {
+                        throw new RuntimeException(error);
+                    }, () -> removeInstanceObserved(instance));
+            removeInstanceObserved(instance);
```

Comment:
Why is observed instance removed here? I would expect the instance to be removed in previous line.

Reply:
As the subscription is blocking (...), I remove the instance when it is finished.
I updated the way it is managed, catching the InterruptedException and do the remove in a finally clause.

Reply:
And the "completed" is almost never called as observe logs is blocking and almost never stopped ...

Reply:
ok